File: day01/solution.py
import sys


def get_calibration_value(data: str) -> int:
    first_number = next(int(c) for c in data if c.isnumeric())
    last_number = next(int(c) for c in reversed(data) if c.isnumeric())
    return first_number*10 + last_number


def main():
    if len(sys.argv) != 2:
        print('Usage: python solution.py <input_file>')
        print('Example: python solution.py input.txt')
        sys.exit(1)

    answer = 0
    in_file = sys.argv[1]
    with open(in_file, 'r') as f:
        for line in f:
            answer += get_calibration_value(line)
    print(answer)


if __name__ == '__main__':
    main()

File: day01/solution.rs
use std::env;
use std::fs::File;
use std::io::{self, prelude::*, BufReader};

fn get_calibration_value(input: &str) -> u32 {
    let first_number = input.chars().find_map(|c| c.to_digit(10)).unwrap();
    let last_number = input.chars().rev().find_map(|c| c.to_digit(10)).unwrap();
    first_number * 10 + last_number
}

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        println!("Usage: {} <input_file>", args[0]);
        println!("Example: {} input.txt", args[0]);
    } else {
        let input_file = &args[1];

        let file = File::open(input_file)?;
        let reader = BufReader::new(file);

        let answer = reader
            .lines()
            .map(|line| line.unwrap())
            .map(|line| get_calibration_value(&line))
            .sum::<u32>();
        println!("{}", answer);
    }
    Ok(())
}
